### DISCUSSION TITLE
Error with raster CHM for locate_trees()

### DISCUSSION THREAD

**Discussion by StephanLo** [r-lidar/lidR#541] [Q&A] [answered]
Hello! 

I'm using lidR 4.0.0 on Windows. 

when giving a raster CHM (as a terra `SpatRaster`) to the `locate_trees()` function it returns the error
> `Error: On-disk rasters not supported without a bbox in 'raster_as_las()'`

I can't find the `raster_as_las()` function, in documentation or in the source code (but I might be looking in wrong places) 

The raster has these attributes:

```r
class       : SpatRaster 
dimensions  : 2125, 1668, 1  (nrow, ncol, nlyr)
resolution  : 0.3, 0.3  (x, y)
extent      : 601523.7, 602024.1, 4912952, 4913590  (xmin, xmax, ymin, ymax)
coord. ref. : WGS 84 / UTM zone 54N (EPSG:32654) 
source      : jinjayama_chm.tif 
name        : jinjayama_chm 
min value   :        -0.077 
max value   :        39.361
```

Any advice would be appreciated, thank you!

**Comment by Jean-Romain** [comment]
Please report a minimal reproducible example.

**Comment by StephanLo** [comment]
Thank you for responding.
It looks like the error happens when the raster is loaded from disk, and is not in memory. 
minimal example: 

```R
library(lidR)  # v 4.0.0
library(terra) # v 1.5-17

LASfile <- system.file("extdata", "Topography.laz", package="lidR")
las <- readLAS(LASfile)
dtm_tin <- rasterize_terrain(las, res = 1, algorithm = tin())

ncloud <- las - dtm_tin
chm <- rasterize_canopy(ncloud, res = 1, algorithm = p2r())


trees <- locate_trees(chm,lmf(ws =5)) # works as expected


## The error happens when the raster is loaded from disk.
writeRaster(chm, "testraster.tif")

chm_disk <- rast("testraster.tif")
locate_trees(chm_disk,lmf(ws =5))
# Error: On-disk rasters not supported without a bbox in 'raster_as_las()'
 ```

**Comment by Jean-Romain** [comment]
Thank you. The problem corresponds to the expected behavior but I do agree that the error is not the one that should be thrown to users. Something more informative should be displayed.

The idea is that if your raster is not loaded in memory it means that it is too big to be loaded in memory. `lidR` needs a raster in memory and does not force loading on-the-fly (because it assumes it is too big). What I forget however is that `terra` does not load in memory by default even for small rasters. Thus:

```r
chm_disk <- rast("testraster.tif")
locate_trees(chm_disk,lmf(ws =5))
```

Looks perfectly fair but is understood by `lidR` as *the raster is too big, abort*. I will improve that. On your side the problem is easy to solve: load the raster in memory. In `raster` their is `load_all()`. I think the equivalent in `terra` is `chm_memory <- chm_disk * 1`

**Comment by Jean-Romain** [comment]
The error is now 

> On-disk rasters are not supported by locate_tree. Use an in-memory raster

**Comment by Jean-Romain** [comment, marked as answer]
If the raster is small enough it is loaded on the fly. Your example now works

**Reply by StephanLo** [reply]
Thank you very much. And also, thank you for developing the package. It is really useful
